// zcatalyst-sdk-node/lib/job-scheduling/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobScheduling = void 0;
const api_request_1 = require("../utils/api-request");
const validator_1 = require("../utils/validator");
const constants_1 = __importDefault(require("../utils/constants"));
const constants_2 = __importDefault(require("./constants"));
const job_1 = __importDefault(require("./job"));
const error_1 = __importDefault(require("./error"));
const cron_1 = __importDefault(require("./cron"));
const { COMPONENT, REQ_METHOD, CREDENTIAL_USER } = constants_1.default;
const { API } = constants_2.default;
class JobScheduling {
    constructor(app) {
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
        this.CRON = new cron_1.default(this.requester);
        this.JOB = new job_1.default(this.requester);
    }
    getComponentName() {
        return COMPONENT.job_scheduling;
    }
    // Jobpool
    /**
     * Get Jobpool details of the specific Jobpool
     * @param jobpoolId - Id of the Jobpool
     * @returns Jobpool details
     */
    getJobpool(jobpoolId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyStringOrNumber)(jobpoolId, 'jobpool_id', true);
            }, error_1.default);
            const request = {
                method: REQ_METHOD.get,
                path: API.JOBPOOL + '/' + jobpoolId,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    /**
     * Get all jobpool details
     * @returns Array of jobpool details
     */
    getAllJobpool() {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                method: REQ_METHOD.get,
                path: API.JOBPOOL,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.JobScheduling = JobScheduling;
