// @zcatalyst/transport/dist-es/utils/constants.js
export const HTTP_CODE_MAP = {
    OK: {
        CODE: 200,
        TEXT: 'OK'
    },
    NO_CONTENT: {
        CODE: 204,
        TEXT: 'NO CONTENT'
    },
    INTERNAL_SERVER_ERROR: {
        CODE: 500,
        TEXT: 'INTERNAL SERVER ERROR'
    },
    RESOURCE_NOT_FOUND: {
        CODE: 404,
        TEXT: 'RESOURCE NOT FOUND'
    },
    BAD_REQUEST: {
        CODE: 400,
        TEXT: 'BAD REQUEST'
    },
    UNAUTHORIZED: {
        CODE: 401,
        TEXT: 'UNAUTHORIZED'
    },
    CONFLICT: {
        CODE: 409,
        TEXT: 'CONFLICT, MAY BE YOU ARE TRYING TO CREATE A RESOURCE THAT ALREADY EXIST!'
    },
    FORBIDDEN: {
        CODE: 403,
        TEXT: 'FORBIDDEN, MAY BE YOUR USAGE REACHED MAX ALLOWED LIMIT!'
    },
    UNEXPECTED: {
        CODE: 101,
        TEXT: 'UNEXPECTED RESPONSE FOUND'
    }
};
export const HTTP_HEADER_MAP = {
    CONTENT_JSON: 'application/json; charset=utf-8',
    AUTHORIZATION_KEY: 'Authorization'
};
export const REQUEST_PROPERTY = {
    BLOB: 'blob'
};
export const X_ZCSRF_TOKEN = 'X-ZCSRF-TOKEN';
export const ZD_CSRPARAM = 'zd_csrparam';
const HTTP_CODE_REV_MAP = {};
const HTTP_CODE_MAP_KEYS = Object.keys(HTTP_CODE_MAP);
for (const KEY of HTTP_CODE_MAP_KEYS)
    HTTP_CODE_REV_MAP[HTTP_CODE_MAP[KEY].CODE] = HTTP_CODE_MAP[KEY];
export { HTTP_CODE_REV_MAP };
