// @zcatalyst/auth-client/dist-cjs/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROJECT = exports.API_DOMAIN = exports.ACCOUNTS_PORTAL_DOMAIN = exports.IS_APPSAIL = exports.ZAID = exports.PROJECT_ID = exports.COMMONPOOL_NAME = exports.CSRF_TOKEN_KEY = exports.APP_URL_DIVIDER = exports.REQUIREMENT = exports.AUTH_STATIC_FILES = exports.AUTH_ERROR_MSG = exports.signUpProperty = exports.UM_QUERY_STRING = exports.UM_PROPERTY = exports.URL_DIVIDER = exports.UM_URL_DIVIDER = void 0;
exports.UM_URL_DIVIDER = {
    PROJECT_USER: 'project-user',
    DISABLE: 'disable',
    ENABLE: 'enable',
    FORGOT_PASSWORD: 'forgotpassword',
    CURRENT: 'current'
};
exports.URL_DIVIDER = {
    RESERVED_URL: '__catalyst',
    AUTH: 'auth',
    PUBLIC_SIGNUP: 'public-signup',
    CHANGE_PASSWORD: 'change-password',
    LOGIN: 'login'
};
exports.UM_PROPERTY = {
    ZAID: 'zaid',
    USER_DETAILS: 'user_details',
    EMAIL_ID: 'email_id',
    NAME: 'name',
    ZAAID: 'zaaid',
    FIRST_NAME: 'first_name',
    LAST_NAME: 'last_name',
    PLATFORM: 'platform_type',
    REDIRECT_URL: 'redirect_url'
};
exports.UM_QUERY_STRING = {
    EMAIL_ID: 'emailId'
};
exports.signUpProperty = {
    FIRST_NAME: 'first_name',
    LAST_NAME: 'last_name',
    EMAIL_ID: 'email_id',
    REDIRECT_URL: 'redirect_url',
    PLATFORM: 'platform_type',
    USER_DETAILS: 'user_details',
    ZAID: 'zaid'
};
exports.AUTH_ERROR_MSG = {
    emptyEmailAddress: 'Please enter your email address',
    noAccountIncludes: 'account does not exist',
    noAccountMsg: 'This account does not exist'
};
exports.AUTH_STATIC_FILES = {
    URL: '/__catalyst/auth/static-file',
    SIGNIN: 'embedded_signin.css',
    SIGNIN_WITH_PROVIDERS_ONLY: 'embedded_signin_providers_only.css',
    FORGOT_PWD: 'embedded_password_reset.css'
};
exports.REQUIREMENT = {
    INIT_REQUIRE: ['zaid', 'project_id']
};
exports.APP_URL_DIVIDER = {
    PROJECT: 'project'
};
exports.CSRF_TOKEN_KEY = 'ZD_CSRF_TOKEN';
exports.COMMONPOOL_NAME = 'catalystApp';
exports.PROJECT_ID = 'PROJECT_ID';
exports.ZAID = 'ZAID';
exports.IS_APPSAIL = 'IS_APPSAIL';
exports.ACCOUNTS_PORTAL_DOMAIN = 'AUTH_DOMAIN';
exports.API_DOMAIN = 'API_DOMAIN';
exports.PROJECT = 'project';
