// @zcatalyst/transport/dist-cjs/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTP_CODE_REV_MAP = exports.ZD_CSRPARAM = exports.X_ZCSRF_TOKEN = exports.REQUEST_PROPERTY = exports.HTTP_HEADER_MAP = exports.HTTP_CODE_MAP = void 0;
exports.HTTP_CODE_MAP = {
    OK: {
        CODE: 200,
        TEXT: 'OK'
    },
    NO_CONTENT: {
        CODE: 204,
        TEXT: 'NO CONTENT'
    },
    INTERNAL_SERVER_ERROR: {
        CODE: 500,
        TEXT: 'INTERNAL SERVER ERROR'
    },
    RESOURCE_NOT_FOUND: {
        CODE: 404,
        TEXT: 'RESOURCE NOT FOUND'
    },
    BAD_REQUEST: {
        CODE: 400,
        TEXT: 'BAD REQUEST'
    },
    UNAUTHORIZED: {
        CODE: 401,
        TEXT: 'UNAUTHORIZED'
    },
    CONFLICT: {
        CODE: 409,
        TEXT: 'CONFLICT, MAY BE YOU ARE TRYING TO CREATE A RESOURCE THAT ALREADY EXIST!'
    },
    FORBIDDEN: {
        CODE: 403,
        TEXT: 'FORBIDDEN, MAY BE YOUR USAGE REACHED MAX ALLOWED LIMIT!'
    },
    UNEXPECTED: {
        CODE: 101,
        TEXT: 'UNEXPECTED RESPONSE FOUND'
    }
};
exports.HTTP_HEADER_MAP = {
    CONTENT_JSON: 'application/json; charset=utf-8',
    AUTHORIZATION_KEY: 'Authorization'
};
exports.REQUEST_PROPERTY = {
    BLOB: 'blob'
};
exports.X_ZCSRF_TOKEN = 'X-ZCSRF-TOKEN';
exports.ZD_CSRPARAM = 'zd_csrparam';
const HTTP_CODE_REV_MAP = {};
exports.HTTP_CODE_REV_MAP = HTTP_CODE_REV_MAP;
const HTTP_CODE_MAP_KEYS = Object.keys(exports.HTTP_CODE_MAP);
for (const KEY of HTTP_CODE_MAP_KEYS)
    HTTP_CODE_REV_MAP[exports.HTTP_CODE_MAP[KEY].CODE] = exports.HTTP_CODE_MAP[KEY];
